src/EditorFeatures/Core/Implementation/AddImports/AddImportsPasteCommandHandler.cs

```diff
@@ -0,0 +1,120 @@
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AddImportsPasteCommandHandler(IThreadingContext threadingContext)
+            => _threadingContext = threadingContext;
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Capture the pre-paste caret position
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretPosition.HasValue)
+            {
+                return;
+            }
+
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            // Don't perform work if we're inside the interactive window
+            if (args.TextView.IsBufferInInteractiveWindow(args.SubjectBuffer))
+            {
+                return;
+            }
+
+            if (!args.SubjectBuffer.GetFeatureOnOffOption(FeatureOnOffOptions.AddImportsOnPaste))
+            {
+                return;
+            }
+
+            // Create a tracking span from the pre-paste caret position that will grow as text is inserted.
+            var trackingSpan = caretPosition.Value.Snapshot.CreateTrackingSpan(caretPosition.Value.Position, 0, SpanTrackingMode.EdgeInclusive);
+
+            // Applying the post-paste snapshot to the tracking span gives us the span of pasted text.
+            var snapshotSpan = trackingSpan.GetSpan(args.SubjectBuffer.CurrentSnapshot);
+            var textSpan = TextSpan.FromBounds(snapshotSpan.Start, snapshotSpan.End);
+
+            AddMissingImportsForPaste(args, executionContext, textSpan);
+        }
+
+        public CommandState GetCommandState(PasteCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
+
+        private void AddMissingImportsForPaste(PasteCommandArgs args, CommandExecutionContext executionContext, TextSpan textSpan)
+        {
+            var sourceTextContainer = args.SubjectBuffer.AsTextContainer();
+            if (!Workspace.TryGetWorkspace(sourceTextContainer, out var workspace))
+            {
+                return;
+            }
+
+            var document = sourceTextContainer.GetOpenDocumentInCurrentContext();
+            if (document is null)
+            {
+                return;
+            }
+
+            using var _ = executionContext.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Adding_missing_import_directives);
+            var cancellationToken = executionContext.OperationContext.UserCancellationToken;
+
+#pragma warning disable VSTHRD102 // Implement internal logic asynchronously
+            var updatedDocument = _threadingContext.JoinableTaskFactory.Run(() => AddMissingImportsAsync(document, textSpan, cancellationToken));
```

Comment:
it looks like we're still blockign the user.  can we not do that?  i'm still not getting how that's valuable.

Reply:
https://github.com/dotnet/roslyn/pull/48501#discussion_r508030664